[quant][graphmode] Get scalar type from observer module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26390 [quant][graph] PyTorch Graph Mode Quantization API
* #26389 [quant][graph] Add inplace argument to InsertObservers and InsertQuantDeQuant
* **#26425 [quant][graphmode] Get scalar type from observer module**

Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17504459](https://our.internmc.facebook.com/intern/diff/D17504459)

jerryzh168 committed:
[quant][graphmode] Get scalar type from observer module

Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Get scalar type from observer module"


Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Get scalar type from observer module"


Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Get scalar type from observer module"


Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Get scalar type from observer module"


Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -178,10 +178,10 @@ Node* insertObserver(
     script::Module& module,
-    const QConfig& qconfig) {
+    const QConfig& qconfig,
+    const std::unordered_set<Value*>& weight_values) {
```
Might be worth to refactor all that into a helper class to avoid passing these sets in every function. Later we'll also add a similar set for 'bias' values.

jerryzh168 replied:
OK will do refactor in a separate PR

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -242,12 +242,27 @@ void fillQConfigMap(
+    Node* n,
+    const script::Module& module,
+    const std::string& module_method_name,
+    std::unordered_set<Value*> weight_values) {
```
Did you mean to pass this by reference?

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -596,7 +622,8 @@ TORCH_API void InsertObservers(
   std::unordered_set<Value*> values_to_skip;
-  InsertObserversImpl(module, method_name, module_qconfig_map, values_to_skip);
+  std::unordered_set<Value*> weight_values;
+  InsertObserversImpl(module, method_name, module_qconfig_map, values_to_skip, weight_values);
```
clang-format :-p

ZolotukhinM reviewed (changes requested):
Could you please add a test for this change? Ideally, the test should fail before this change and pass after (I also suspect that such test would catch that you're not actually passing the weight set by reference). Otherwise the code looks good!

jerryzh168 committed:
Update on "[quant][graphmode] Get scalar type from observer module"


Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Get scalar type from observer module"


Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -304,32 +319,52 @@ void InsertObserversImpl(
           // the child module.
           auto module_instance = v->node()->inputs()[0];
           auto module_method_name = v->node()->s(attr::name);
+          // TODO: looks like this block is not related to v? maybe we should
```
The code below uses `v-node()`. I'm not sure this TODO comment is correct.

jerryzh168 replied:
I mean we don't need `v`, we just need `node`

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -242,12 +242,28 @@ void fillQConfigMap(
+    const std::string& module_method_name,
+    std::unordered_set<Value*>& weight_values) {
+  auto method = module.get_method(module_method_name);
+  auto graph = method.graph();
```
We can just pass graph here instead of `module` and `method_name`.

ZolotukhinM reviewed (approved):
Overall looks good! Please fix a couple of remarks inline before landing.

jerryzh168 commented on test/test_jit.py:
```diff
@@ -1061,11 +1064,54 @@ def test_module(module, relu_call, num_observers):
+                weight=weight_observer._c)
+        }
+        torch._C._jit_pass_insert_observers(m._c, "forward", qconfig_dict, True)
+        assert m._c._get_module('conv')._get_module('observer_for_input.1')._get_attribute('dtype') == 13  # torch.quint8
```
we can but they should be pretty stable, I don't expect we'll be deleting things from this list: https://github.com/pytorch/pytorch/blob/master/c10/core/ScalarType.h#L36

ZolotukhinM replied:
I think my main concern is that they are internal details, and thus we shouldn't rely on them in our tests.

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -304,32 +319,52 @@ void InsertObserversImpl(
                 "or child instance in insert_observers_pass right now");
+            std::cout << "propagate values for " << module_method_name << std::endl;
+            std::cout << "weight values: " << weight_values.size() << std::endl;
+            propagateWeightValues(
```
OK

jerryzh168 committed:
Update on "[quant][graphmode] Get scalar type from observer module"


Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
OK. will change them

Sent from my iPhone

On Sep 20, 2019, at 12:19, Mikhail Zolotukhin <notifications@github.com> wrote:

﻿

@ZolotukhinM commented on this pull request.

________________________________

In test/test_jit.py<https://github.com/pytorch/pytorch/pull/26425#discussion_r326768590>:

> +                return F.relu(self.conv(x))
+
+        def get_forward(m):
+            return m._c._get_method("forward")
+
+        m = torch.jit.script(M())
+        observer = torch.jit.script(Observer())
+        weight_observer = torch.jit.script(WeightObserver())
+        qconfig_dict = {
+            '':
+            QConfig(
+                activation=observer._c,
+                weight=weight_observer._c)
+        }
+        torch._C._jit_pass_insert_observers(m._c, "forward", qconfig_dict, True)
+        assert m._c._get_module('conv')._get_module('observer_for_input.1')._get_attribute('dtype') == 13  # torch.quint8


I think my main concern is that they are internal details, and thus we shouldn't rely on them in our tests.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/26425?email_source=notifications&email_token=ABF2R2KV6YHGZBJACLXY623QKUOYLA5CNFSM4IYDO2L2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFO2TSQ#discussion_r326768590>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABF2R2NLMSWAAORE5OGJZP3QKUOYLANCNFSM4IYDO2LQ>.

jerryzh168 committed:
Update on "[quant][graphmode] Get scalar type from observer module"


Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17504459](https://our.internmc.facebook.com/intern/diff/D17504459)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Get scalar type from observer module"


Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17504459](https://our.internmc.facebook.com/intern/diff/D17504459)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Get scalar type from observer module"


Summary:
Currently the scalar type is hardcoded for weight and normal tensor
but what we want is to get it from corresponding observer module

Test Plan:
there are some known issues right now,
will test e2e later when all the issues are fixed

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17504459](https://our.internmc.facebook.com/intern/diff/D17504459)

[ghstack-poisoned]
